openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,428 @@
+
+    if wav[pos:pos+2] != b'\x01\x00':
+        return "Samples not in PCM-format."
+
+    inopt['channels'] = int.from_bytes(wav[pos+2:pos+4], le)
+    if inopt['channels'] not in (1, 2):
+        return "WAVE-file has {} channels. Need 1 or 2."\
+                .format(inopt['channels'])
+
+    inopt['input_rate'] = int.from_bytes(wav[pos+4:pos+8], le)
+    if inopt['input_rate'] > 48000:
+        return "Inputrate {} too high for opus."\
+                .format(inopt['input_rate'])
+
+    bytes_per_second = int.from_bytes(wav[pos+8:pos+12], le)
+    # wframe_len is the space needed to save one sample (padded to whole bytes)
+    # for each channel.
+    wframe_len = int.from_bytes(wav[pos+12:pos+14], le)
+    bits_per_sample = int.from_bytes(wav[pos+14:pos+16], le)
+    if bits_per_sample not in range(1,17):
+        return "Samplesize is {}, but should be between 1 and 16 bits inclusive."\
+                .format(bits_per_sample)
+
+    assert(wframe_len * inopt['input_rate'] == bytes_per_second)
+    assert((bits_per_sample + 7) // 8 * inopt['channels'] == wframe_len)
+
+    inopt['bits_per_sample'] = bits_per_sample
+    inopt['wframe_len'] = wframe_len
+
+    pos += chk_len + (chk_len & 1)
+    # Search for 'data' chunk.
+    while wav[pos:pos+4] != b'data':
+        chk_len = int.from_bytes(wav[pos+4:pos+8], le)
+        pos += 8 + chk_len + (chk_len & 1)  # Pad to even byte offset.
+        if pos >= len(wav):
+            return "Could not find 'data'-chunk in Wave-file."
+
+    chk_len = int.from_bytes(wav[pos+4:pos+8], le)
+    pos += 8
+    if len(wav) - pos < chk_len:
+        err = "Stated data-length {} in WAVE-header too high. " +\
+                "It should be at most {}. Is the file maybe truncated?"
+        return err.format(chk_len, len(wav) - pos)
+
+    inopt['header_len'] = pos
+    inopt['trailer_len'] = len(wav) - pos - chk_len
+    return inopt
+
+
+cdef size_t write_ogg_page(size_t offset, bytearray outdata, ogg.ogg_page *og):
+    '''
+    Write og at offset into outdata.
+    If succesful, returns new offset, else 0.
+    '''
+    if PyByteArray_GET_SIZE(outdata) < offset + og.header_len + og.body_len:
+        try:
+            PyByteArray_Resize(outdata, offset + og.header_len + og.body_len)
+            # TODO increase more than necessary
+        except MemoryError:
+            return 0
+
+    cdef char *out = PyByteArray_AS_STRING(outdata) + offset
+    memcpy(out, og.header, og.header_len)
+    memcpy(out + og.header_len, og.body, og.body_len)
+    return offset + og.header_len + og.body_len
+
+
+cdef void free_buffers(ogg.ogg_stream_state *os, opus.OpusEncoder *oe,
+                       unsigned char *pymem1, unsigned char *pymem2):
+    ogg.ogg_stream_clear(os)
+    if oe != NULL:
+        opus.opus_encoder_destroy(oe)
+    if pymem1 != NULL:
+        PyMem_Free(pymem1)
+    if pymem2 != NULL:
+        PyMem_Free(pymem2)
+    return
+
+
+cdef upsample(indata, dict opts, int target_rate):
```

Comment:
i think you can speed up things if you declare the `indata` type or use a memoryview.